### PR TITLE
Support import headers and include to metal_library rule

### PR DESCRIPTION
**Pull request by vikage**
This pull request support import headers and include to metal_library rule. Fix will fix issue #39  and #40 and fix others compile issue

### CODE REVIEW

**Comment by vikage** [comment]
@thii Please help me review this pull request

File: apple/metal_library.bzl [hunk 1]
```diff
@@ -22,55 +22,58 @@ load(
 load("@bazel_skylib//lib:dicts.bzl", "dicts")
 load("@bazel_skylib//lib:paths.bzl", "paths")
 load(":common.bzl", "METAL_FILE_TYPES")
+load("@build_bazel_rules_apple//apple/internal:resources.bzl","resources")
```
Comment (giuk-jung) [inline apple/metal_library.bzl]: Please put a space after the comma.

File: apple/metal_library.bzl [hunk 2]
```diff
@@ -114,12 +129,28 @@ A list of `.metal` source files that will be compiled into the library.
 An output `.metallib` filename. Defaults to `default.metallib` if unspecified.
 """,
         ),
+        "minimum_os_version": attr.string(
+            mandatory = True,
```
Comment (giuk-jung) [inline apple/metal_library.bzl]: For `minimum_os_version`, please use `DEFAULT_MINIMUM_OS_VERSION` in apple/common.bzl by default

File: apple/metal_library.bzl [hunk 3]
```diff
@@ -22,55 +22,58 @@ load(
...
-
+        
         args = [
             "metal",
             "-c",
             "-target",
             target,
-            "-ffast-math",
-            "-o",
+            "-ffast-math"]
```
Comment (giuk-jung) [inline apple/metal_library.bzl]:
```suggestion
            "-ffast-math"
        ]
```

File: apple/metal_library.bzl [hunk 4]
```diff
@@ -22,55 +22,58 @@ load(
...
+    for includePath in ctx.attr.includes:
+        includes_input.append("-I{}".format(includePath))
+    
     for input_metal in ctx.files.srcs:
         air_file = ctx.actions.declare_file(
             paths.replace_extension(input_metal.basename, ".air"),
         )
         air_files.append(air_file)
-
+
```
Comment (giuk-jung) [inline apple/metal_library.bzl]:
```suggestion

```

File: apple/metal_library.bzl [hunk 5]
```diff
@@ -97,7 +100,13 @@ def _metal_library_impl(ctx):
...
     )
 
-    return [DefaultInfo(files = depset([output_metallib]))]
+    resource_provider = resources.bucketize_typed(
+        [output_metallib],
+        bucket_type = "unprocessed",
+    )
+    return [
+        resource_provider,
+    ]
```
Comment (giuk-jung) [inline apple/metal_library.bzl]:
```suggestion
    # Return the provider for the new bundling logic of rules_apple.
    return [
         resources.bucketize_typed([output_metallib], "unprocessed"),
    ]
```

File: apple/metal_library.bzl [hunk 6]
```diff
@@ -106,6 +115,12 @@ metal_library = rule(
             allow_files = METAL_FILE_TYPES,
             doc = """\
 A list of `.metal` source files that will be compiled into the library.
+""",
+        ),
+        "hdrs": attr.label_list(
```
Comment (giuk-jung) [inline apple/metal_library.bzl:139]: Isn't it necessary to make `hdrs` mandatory?

File: apple/metal_library.bzl [hunk 7]
```diff
@@ -129,4 +160,4 @@ load("@rules_apple_line//apple:metal_library.bzl", "metal_library")
     outputs = {
         "out": "%{out}",
     },
-)
+)
```
Comment (giuk-jung) [inline apple/metal_library.bzl]: Please insert a new line.

File: apple/metal_library.bzl (hunk 6 above)
Comment (vikage) [reply apple/metal_library.bzl:139]: No. hdrs don't need mandatory. Because not all metal file need hdrs

**Comment by vikage** [comment]
@giuk-jung Please help me check the update

File: apple/metal_library.bzl [hunk 8]
```diff
@@ -97,7 +101,10 @@ def _metal_library_impl(ctx):
         mnemonic = "MetallibCompile",
     )
 
-    return [DefaultInfo(files = depset([output_metallib]))]
+    # Return the provider for the new bundling logic of rules_apple.
+    return [
+         resources.bucketize_typed([output_metallib], "unprocessed"),
+    ]
```
Comment (giuk-jung) [inline apple/metal_library.bzl]:
```suggestion
    return [
        DefaultInfo(
            files = depset([output_metallib]),
        ),
        resources.bucketize_typed([output_metallib], "unprocessed"),
    ]
```

**Comment by giuk-jung** [comment]
@vikage , Thanks so much for the contribution! 
I just had one last suggestion.

File: apple/metal_library.bzl (hunk 2 above)
Comment (thii) [reply apple/metal_library.bzl]: This should come from the configuration transition.

File: apple/metal_library.bzl [hunk 9]
```diff
@@ -21,35 +21,36 @@ load(
...
-        platform_string = "macosx"
-
+    platform = ctx.fragments.apple.single_arch_platform
     environment = "" if platform.is_device else "-simulator"
+    target_os_version = ctx.attr.minimum_os_version
 
     return "air64-apple-{platform}{version}{environment}".format(
         environment = environment,
-        platform = platform_string,
+        platform = ctx.attr.platform,
```
Comment (thii) [inline apple/metal_library.bzl]: What's the reason for this change?

Comment (vikage) [reply apple/metal_library.bzl]: @thii  Current, `platform_string` is undefined when you build for ios and the line `platform_prerequisites = platform_support.platform_prerequisites_from_rule_ctx(ctx)` show error when run( i mention in issue #40)

File: apple/metal_library.bzl (hunk 2 above)
Comment (vikage) [reply apple/metal_library.bzl]: @thii I changed to use `DEFAULT_MINIMUM_OS_VERSION`. I don't understand your comment

Comment (giuk-jung) [reply apple/metal_library.bzl]: @thii , I don't understand about configuration transition either, but are you expecting something like this?
```
platform = apple_fragment.single_arch_platform
if not minimum_os_version:
    minimum_os_version = ctx.attr._xcode_config[apple_common.XcodeVersionConfig].minimum_os_for_platform_type(
        platform.platform_type,
    )
```

Comment (thii) [reply apple/metal_library.bzl]: If you're building this as part of a top-level bundling target—for example, an `ios_application` target—then the `minimum_os_version` value should be passed down from that `ios_application`'s `minimum_os_version` attribute instead.

Comment (vikage) [reply apple/metal_library.bzl]: @thii Can you give me some hint to do that. I look at ctx but it do not contain that info. And why `swift_static_framework` and `mixed_static_framework` rules still need `minimum_os_version` while we can pass down `minimum_os_version`.

File: apple/metal_library.bzl (hunk 9 above)
Comment (thii) [reply apple/metal_library.bzl]: This is another bug. I'll take a look.

Comment (thii) [reply apple/metal_library.bzl]: https://github.com/line/rules_apple_line/pull/42

File: apple/metal_library.bzl (hunk 2 above)
Comment (thii) [reply apple/metal_library.bzl]: You can get it from `platform_prerequisites.minimum_os`.

> why swift_static_framework and mixed_static_framework rules still need minimum_os_version while we can pass down minimum_os_version.

These are also top-level bundling rules.

Comment (vikage) [reply apple/metal_library.bzl]: @thii I looking at current code, It get `platform_prerequisites` from ctx. But throw error(#40)

Comment (vikage) [reply apple/metal_library.bzl]: I found you just open another pull request to fix `platform_prerequisites` problem. I will change to use `platform_prerequisites` when it work again.

Comment (thii) [reply apple/metal_library.bzl]: Merged the other PR. Please rebase your branch.

**Comment by vikage** [comment]
@thii Please help me review this pull request. I rebased the latest code.

**Review by thii — APPROVED** [review]
Thanks @vikage!
